reflect.go

```diff
@@ -0,0 +1,387 @@
+	}
+}
+
+// recordArgReflected finds the type(s) of a function argument, which is being used in reflection
+// and excludes these types from obfuscation
+// It also checks if this argument has any relation to a function paramter and returns it if found.
+func (tf *transformer) recordArgReflected(val ssa.Value, visited map[ssa.Value]bool) *ssa.Parameter {
+	// make sure we visit every val only once, otherwise there will be infinite recursion
+	if visited[val] {
+		return nil
+	}
+
+	/* fmt.Printf("val: %v %T %v\n", val, val, val.Type()) */
+	visited[val] = true
+
+	switch val := val.(type) {
+	case *ssa.IndexAddr:
+		for _, ref := range *val.Referrers() {
+			if store, ok := ref.(*ssa.Store); ok {
+				tf.recordArgReflected(store.Val, visited)
+			}
+		}
+		return tf.recordArgReflected(val.X, visited)
+	case *ssa.Slice:
+		return tf.recordArgReflected(val.X, visited)
+	case *ssa.MakeInterface:
+		return tf.recordArgReflected(val.X, visited)
+	case *ssa.UnOp:
+		return tf.recordArgReflected(val.X, visited)
+	case *ssa.FieldAddr:
+		return tf.recordArgReflected(val.X, visited)
+
+	case *ssa.Alloc:
+		/* fmt.Printf("recording val %v \n", *val.Referrers()) */
+		tf.recursivelyRecordAsNotObfuscated(val.Type())
+
+		for _, ref := range *val.Referrers() {
+			if idx, ok := ref.(*ssa.IndexAddr); ok {
+				tf.recordArgReflected(idx, visited)
+			}
+		}
+
+		// relatedParam needs to revisit nodes so create an empty map
+		visited := make(map[ssa.Value]bool)
+
+		// check if the found alloc gets tainted by function parameters
+		return relatedParam(val, visited)
+
+	case *ssa.Const:
+		tf.recursivelyRecordAsNotObfuscated(val.Type())
+	case *ssa.Global:
+		tf.recursivelyRecordAsNotObfuscated(val.Type())
+
+		// TODO: this might need similar logic to *ssa.Alloc, however
+		// reassigning a function param to a global variable and then reflecting
+		// it is probably unlikely to occur
+	case *ssa.Parameter:
+		// this only finds the parameters who want to be found,
+		// otherwise relatedParam is used for more in depth analysis
+		return val
+	}
+
+	return nil
+}
+
+// relatedParam checks if a route to a function paramter can be constructed
+// from a ssa.Value, and returns the paramter if it found one.
+func relatedParam(val ssa.Value, visited map[ssa.Value]bool) *ssa.Parameter {
+	// every val should only be visited once to prevent infinite recursion
+	if visited[val] {
+		return nil
+	}
+
+	/* fmt.Printf("related val: %v %T %v\n", val, val, val.Type()) */
+
+	visited[val] = true
+
+	if param, ok := val.(*ssa.Parameter); ok {
+		// a paramter has been found
+		return param
+	}
+
+	if unop, ok := val.(*ssa.UnOp); ok {
+		if param := relatedParam(unop.X, visited); param != nil {
+			return param
+		}
+	}
+
+	if addr, ok := val.(*ssa.FieldAddr); ok {
+		/* fmt.Printf("addr: %v\n", addr)
+		fmt.Printf("addr.X: %v %T\n", addr.X, addr.X) */
+
+		if param := relatedParam(addr.X, visited); param != nil {
+			return param
+		}
+
```

Comment:
please avoid empty lines at the end of blocks - it just adds vertical space for no benefit, I think. there are a few others in the code. gofumpt deletes these, I believe.